File: apps/web/lib/purchase-orders.ts
import { apiFetch, apiFetchResult } from '@/lib/api';

export type UUID = string;
export type PurchaseOrderStatus =
  | 'DRAFT'
  | 'SUBMITTED'
  | 'APPROVED'
  | 'REJECTED'
  | 'PARTIALLY_RECEIVED'
  | 'RECEIVED'
  | 'CANCELLED'
  | 'CANCELLED_WITH_RECEIPTS';

export const PURCHASE_ORDER_STATUSES: readonly PurchaseOrderStatus[] = [
  'DRAFT',
  'SUBMITTED',
  'APPROVED',
  'REJECTED',
  'PARTIALLY_RECEIVED',
  'RECEIVED',
  'CANCELLED',
  'CANCELLED_WITH_RECEIPTS',
];

export interface DeliveryAddress {
  line1: string | null;
  line2: string | null;
  city: string | null;
  region: string | null;
  postal_code: string | null;
  country_code: string | null;
}

export interface DeliveryAddressInput {
  line1?: string | null;
  line2?: string | null;
  city?: string | null;
  region?: string | null;
  postal_code?: string | null;
  country_code?: string | null;
}

export interface PurchaseOrderLine {
  id: UUID;
  line_number: number;
  inventory_item_id: UUID;
  item_code: string;
  item_name: string;
  item_sku: string | null;
  description: string;
  line_note: string | null;
  ordered_quantity: string;
  ordered_unit: string;
  canonical_unit: string;
  ordered_quantity_canonical: string;
  received_quantity: string;
  received_quantity_canonical: string;
  unit_price: string;
  extended_amount: string;
  created_at: string;
  updated_at: string;
}

export interface PurchaseOrder {
  id: UUID;
  organization_id: UUID;
  farm_id: UUID | null;
  business_partner_id: UUID;
  po_number: string;
  supplier_reference: string | null;
  status: PurchaseOrderStatus;
  currency_code: string;
  order_date: string;
  expected_delivery_date: string | null;
  delivery_address: DeliveryAddress | null;
  notes: string | null;
  supplier_code: string;
  supplier_legal_name: string;
  supplier_trading_name: string | null;
  version: number;
  created_by_id: UUID;
  submitted_by_id: UUID | null;
  submitted_at: string | null;
  approved_by_id: UUID | null;
  approved_at: string | null;
  rejected_by_id: UUID | null;
  rejected_at: string | null;
  cancelled_by_id: UUID | null;
  cancelled_at: string | null;
  created_at: string;
  updated_at: string;
  subtotal: string;
  lines: PurchaseOrderLine[];
}

export interface PurchaseOrderPage {
  items: PurchaseOrder[];
  next_cursor: string | null;
}

export interface PurchaseOrderTransition {
  id: UUID;
  purchase_order_id: UUID;
  actor_id: UUID;
  from_status: PurchaseOrderStatus | null;
  to_status: PurchaseOrderStatus;
  operation: string;
  reason: string | null;
  occurred_at: string;
}

export interface PurchaseOrderTransitionPage {
  items: PurchaseOrderTransition[];
  next_cursor: string | null;
}

export interface PurchaseOrderLineInput {
  inventory_item_id: UUID;
  ordered_quantity: string;
  ordered_unit: string;
  unit_price: string;
  description?: string | null;
  line_note?: string | null;
}

export interface PurchaseOrderUpdateLineInput extends PurchaseOrderLineInput {
  id?: UUID;
}

export interface CreatePurchaseOrderBody {
  business_partner_id: UUID;
  currency_code: string;
  order_date: string;
  expected_delivery_date?: string | null;
  delivery_address?: DeliveryAddressInput | null;
  supplier_reference?: string | null;
  notes?: string | null;
  farm_id?: UUID | null;
  lines?: PurchaseOrderLineInput[];
}

export interface UpdatePurchaseOrderBody {
  expected_version: number;
  business_partner_id?: UUID;
  currency_code?: string;
  order_date?: string;
  expected_delivery_date?: string | null;
  delivery_address?: DeliveryAddressInput | null;
  supplier_reference?: string | null;
  notes?: string | null;
  farm_id?: UUID | null;
  lines?: PurchaseOrderUpdateLineInput[];
}

export interface ListPurchaseOrdersParams {
  organizationId: UUID;
  farmId?: UUID;
  businessPartnerId?: UUID;
  statuses?: readonly PurchaseOrderStatus[];
  orderDateFrom?: string;
  orderDateTo?: string;
  expectedDeliveryFrom?: string;
  expectedDeliveryTo?: string;
  search?: string;
  cursor?: string;
  limit?: number;
  signal?: AbortSignal;
}

export function purchaseOrderListPath(params: ListPurchaseOrdersParams): string {
  const query = new URLSearchParams();
  if (params.farmId) query.set('farm_id', params.farmId);
  if (params.businessPartnerId) query.set('business_partner_id', params.businessPartnerId);
  for (const status of params.statuses ?? []) query.append('status', status);
  if (params.orderDateFrom) query.set('order_date_from', params.orderDateFrom);
  if (params.orderDateTo) query.set('order_date_to', params.orderDateTo);
  if (params.expectedDeliveryFrom) query.set('expected_delivery_from', params.expectedDeliveryFrom);
  if (params.expectedDeliveryTo) query.set('expected_delivery_to', params.expectedDeliveryTo);
  if (params.search) query.set('search', params.search);
  if (params.cursor) query.set('cursor', params.cursor);
  const limit = Math.min(200, Math.max(1, params.limit ?? 50));
  query.set('limit', String(limit));
  return `/v1/organizations/${params.organizationId}/purchase-orders?${query.toString()}`;
}

export async function listPurchaseOrders(
  params: ListPurchaseOrdersParams,
): Promise<PurchaseOrderPage> {
  return apiFetch<PurchaseOrderPage>(purchaseOrderListPath(params), { signal: params.signal });
}

export async function createPurchaseOrder(
  organizationId: UUID,
  body: CreatePurchaseOrderBody,
): Promise<PurchaseOrder> {
  return apiFetch<PurchaseOrder>(`/v1/organizations/${organizationId}/purchase-orders`, {
    method: 'POST',
    body: JSON.stringify(body),
  });
}

export async function getPurchaseOrder(id: UUID, signal?: AbortSignal): Promise<PurchaseOrder> {
  return apiFetch<PurchaseOrder>(`/v1/purchase-orders/${id}`, { signal });
}

export async function updatePurchaseOrder(
  id: UUID,
  body: UpdatePurchaseOrderBody,
): Promise<PurchaseOrder> {
  return apiFetch<PurchaseOrder>(`/v1/purchase-orders/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(body),
  });
}

export interface LifecycleResponse {
  purchaseOrder: PurchaseOrder;
  replayed: boolean;
}

async function lifecycleRequest(
  id: UUID,
  operation: 'submit' | 'withdraw' | 'approve' | 'reject' | 'revise' | 'cancel',
  body?: { reason?: string },
): Promise<LifecycleResponse> {
  const result = await apiFetchResult<PurchaseOrder>(`/v1/purchase-orders/${id}/${operation}`, {
    method: 'POST',
    ...(body ? { body: JSON.stringify(body) } : {}),
  });
  return {
    purchaseOrder: result.data,
    replayed: result.response.headers.get('X-Idempotent-Replay') === 'true',
  };
}

export async function submitPurchaseOrder(id: UUID): Promise<LifecycleResponse> {
  return lifecycleRequest(id, 'submit');
}

export async function withdrawPurchaseOrder(id: UUID, reason: string): Promise<LifecycleResponse> {
  return lifecycleRequest(id, 'withdraw', { reason });
}

export async function approvePurchaseOrder(id: UUID, reason?: string): Promise<LifecycleResponse> {
  return lifecycleRequest(id, 'approve', reason === undefined ? undefined : { reason });
}

export async function rejectPurchaseOrder(id: UUID, reason: string): Promise<LifecycleResponse> {
  return lifecycleRequest(id, 'reject', { reason });
}

export async function revisePurchaseOrder(id: UUID, reason: string): Promise<LifecycleResponse> {
  return lifecycleRequest(id, 'revise', { reason });
}

export async function cancelPurchaseOrder(id: UUID, reason: string): Promise<LifecycleResponse> {
  return lifecycleRequest(id, 'cancel', { reason });
}

export async function listPurchaseOrderTransitions(
  id: UUID,
  params: { cursor?: string; limit?: number; signal?: AbortSignal } = {},
): Promise<PurchaseOrderTransitionPage> {
  const query = new URLSearchParams();
  if (params.cursor) query.set('cursor', params.cursor);
  query.set('limit', String(Math.min(200, Math.max(1, params.limit ?? 50))));
  return apiFetch<PurchaseOrderTransitionPage>(
    `/v1/purchase-orders/${id}/transitions?${query.toString()}`,
    { signal: params.signal },
  );
}

File: apps/api/app/api/v1/endpoints/purchase_orders.py
"""Release 6.0.3 Purchase Order REST endpoints."""

from __future__ import annotations

import uuid
from collections.abc import Awaitable, Callable
from datetime import date
from decimal import ROUND_HALF_UP, Context, Decimal, localcontext
from typing import Annotated

from fastapi import APIRouter, Body, Depends, HTTPException, Query, Response, status
from sqlalchemy import select
from sqlalchemy.orm import selectinload

from app.deps import CurrentOrganization, CurrentUser, DBSession, RequestCtx
from app.models.farm import Farm
from app.models.membership import OrganizationMembership
from app.models.purchase_order import PurchaseOrder, PurchaseOrderStatus
from app.repositories.audit_repo import AuditRepository
from app.repositories.business_partner import BusinessPartnerRepository
from app.repositories.purchase_order import (
    PurchaseOrderLineRepository,
    PurchaseOrderRepository,
    PurchaseOrderSequenceRepository,
    PurchaseOrderTransitionRepository,
)
from app.schemas.purchase_order import (
    ApprovalPayload,
    LifecycleReasonPayload,
    PurchaseOrderCreate,
    PurchaseOrderLineResponse,
    PurchaseOrderPage,
    PurchaseOrderResponse,
    PurchaseOrderTransitionPage,
    PurchaseOrderTransitionResponse,
    PurchaseOrderUpdate,
)
from app.security.authorize import resolve_permission_scopes
from app.services.purchase_order import LifecycleResult, PurchaseOrderService

router = APIRouter(tags=["purchase-orders"])

_RESPONSE_ARITHMETIC_CONTEXT = Context(prec=64, rounding=ROUND_HALF_UP)
_RESPONSE_QUANTUM = Decimal("0.000001")


def _get_service(session: DBSession) -> PurchaseOrderService:
    return PurchaseOrderService(
        po_repo=PurchaseOrderRepository(session),
        line_repo=PurchaseOrderLineRepository(session),
        transition_repo=PurchaseOrderTransitionRepository(session),
        sequence_repo=PurchaseOrderSequenceRepository(session),
        partner_repo=BusinessPartnerRepository(session),
        audit_repo=AuditRepository(session),
    )


ServiceDep = Annotated[PurchaseOrderService, Depends(_get_service)]


def _not_found() -> HTTPException:
    return HTTPException(
        status.HTTP_404_NOT_FOUND,
        {"code": "not_found", "message": "Purchase Order not found.", "context": {}},
    )


def _forbidden(permission: str) -> HTTPException:
    return HTTPException(
        status.HTTP_403_FORBIDDEN,
        f"Missing required permission: {permission}",
    )


async def _applicable_scopes(session: DBSession, user: CurrentUser, po: PurchaseOrder):
    scopes = await resolve_permission_scopes(session, user)
    return [
        scope
        for scope in scopes
        if (scope.organization_id is None and scope.farm_id is None)
        or (scope.organization_id == po.organization_id and scope.farm_id is None)
        or (po.farm_id is not None and scope.farm_id == po.farm_id)
    ]


async def _load_visible_po(
    purchase_order_id: uuid.UUID,
    user: CurrentUser,
    session: DBSession,
    service: ServiceDep,
) -> PurchaseOrder:
    po = await service.po_repo.get_by_id(purchase_order_id)
    if po is None:
        raise _not_found()
    if user.is_superuser:
        return po

    scopes = await resolve_permission_scopes(session, user)
    applicable = await _applicable_scopes(session, user, po)
    if applicable:
        return po

    membership = (
        await session.execute(
            select(OrganizationMembership.id).where(
                OrganizationMembership.user_id == user.id,
                OrganizationMembership.organization_id == po.organization_id,
                OrganizationMembership.is_active.is_(True),
                OrganizationMembership.deleted_at.is_(None),
            )
        )
    ).scalar_one_or_none()
    # A same-tenant user with no grants reaches the explicit 403 permission
    # boundary. A scoped user attempting another farm remains tenant-hidden.
    if membership is not None and not scopes:
        return po
    raise _not_found()


async def _require_po_permission(
    permission: str,
    po: PurchaseOrder,
    user: CurrentUser,
    session: DBSession,
) -> None:
    scopes = await _applicable_scopes(session, user, po)
    if not any("*" in scope.permissions or permission in scope.permissions for scope in scopes):
        raise _forbidden(permission)


def _po_response(po: PurchaseOrder, service: PurchaseOrderService) -> PurchaseOrderResponse:
    lines = []
    for line in sorted(po.lines, key=lambda row: row.line_number):
        with localcontext(_RESPONSE_ARITHMETIC_CONTEXT):
            extended = (Decimal(line.ordered_quantity) * Decimal(line.unit_price)).quantize(
                _RESPONSE_QUANTUM, rounding=ROUND_HALF_UP
            )
        lines.append(
            PurchaseOrderLineResponse.model_validate(
                {
                    **{
                        field: getattr(line, field)
                        for field in PurchaseOrderLineResponse.model_fields
                        if field != "extended_amount"
                    },
                    "extended_amount": extended,
                }
            )
        )
    values = {
        field: getattr(po, field)
        for field in PurchaseOrderResponse.model_fields
        if field not in {"subtotal", "lines"}
    }
    return PurchaseOrderResponse.model_validate(
        {**values, "subtotal": service.subtotal(po), "lines": lines}
    )


async def _fresh_po_response(
    po_id: uuid.UUID, service: PurchaseOrderService
) -> PurchaseOrderResponse:
    po = (
        await service.session.execute(
            select(PurchaseOrder)
            .where(PurchaseOrder.id == po_id)
            .options(selectinload(PurchaseOrder.lines))
            .execution_options(populate_existing=True)
        )
    ).scalar_one()
    return _po_response(po, service)


async def _lifecycle_response(
    result: LifecycleResult,
    response: Response,
    service: PurchaseOrderService,
) -> PurchaseOrderResponse:
    if result.replay:
        response.headers["X-Idempotent-Replay"] = "true"
    return await _fresh_po_response(result.purchase_order.id, service)


@router.get(
    "/organizations/{organization_id}/purchase-orders",
    response_model=PurchaseOrderPage,
)
async def list_purchase_orders(
    organization_id: uuid.UUID,
    _organization: CurrentOrganization,
    user: CurrentUser,
    service: ServiceDep,
    farm_id: uuid.UUID | None = Query(default=None),
    business_partner_id: uuid.UUID | None = Query(default=None),
    statuses: Annotated[list[PurchaseOrderStatus] | None, Query(alias="status")] = None,
    order_date_from: date | None = Query(default=None),
    order_date_to: date | None = Query(default=None),
    expected_delivery_from: date | None = Query(default=None),
    expected_delivery_to: date | None = Query(default=None),
    search: str | None = Query(default=None, max_length=255),
    cursor: str | None = Query(default=None),
    limit: int = Query(default=50, ge=1, le=200),
) -> PurchaseOrderPage:
    del _organization
    scopes = await resolve_permission_scopes(service.session, user)
    org_read = any(
        (scope.organization_id is None or scope.organization_id == organization_id)
        and scope.farm_id is None
        and ("*" in scope.permissions or "purchase_order.read" in scope.permissions)
        for scope in scopes
    )
    farm_ids = sorted(
        {
            scope.farm_id
            for scope in scopes
            if scope.organization_id == organization_id
            and scope.farm_id is not None
            and ("*" in scope.permissions or "purchase_order.read" in scope.permissions)
        },
        key=str,
    )
    if not org_read and not farm_ids:
        raise _forbidden("purchase_order.read")
    if farm_id is not None:
        if not org_read and farm_id not in farm_ids:
            raise _not_found()
        farm_ids = [farm_id]
        org_read = False

    rows, next_cursor = await service.po_repo.list_page(
        organization_id,
        farm_ids=farm_ids,
        org_scope=org_read,
        business_partner_id=business_partner_id,
        statuses=statuses,
        supplier_search=search,
        order_date_from=order_date_from,
        order_date_to=order_date_to,
        expected_delivery_from=expected_delivery_from,
        expected_delivery_to=expected_delivery_to,
        cursor=cursor,
        limit=limit,
    )
    return PurchaseOrderPage(
        items=[_po_response(po, service) for po in rows], next_cursor=next_cursor
    )


@router.post(
    "/organizations/{organization_id}/purchase-orders",
    response_model=PurchaseOrderResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_purchase_order(
    organization_id: uuid.UUID,
    payload: PurchaseOrderCreate,
    _organization: CurrentOrganization,
    user: CurrentUser,
    request_ctx: RequestCtx,
    service: ServiceDep,
) -> PurchaseOrderResponse:
    del _organization
    data = payload.model_dump()
    lines = data.pop("lines")
    if payload.farm_id is not None:
        farm = await service.session.get(Farm, payload.farm_id)
        if (
            farm is None
            or farm.organization_id != organization_id
            or farm.deleted_at is not None
            or not farm.is_active
        ):
            raise _not_found()
        if not user.is_superuser:
            scopes = await resolve_permission_scopes(service.session, user)
            has_org_scope = any(
                scope.organization_id == organization_id and scope.farm_id is None
                for scope in scopes
            )
            has_farm_scope = any(scope.farm_id == payload.farm_id for scope in scopes)
            if not has_org_scope and not has_farm_scope:
                raise _not_found()
    po = await service.create(
        actor=user,
        organization_id=organization_id,
        lines=lines,
        request_ctx=request_ctx,
        **data,
    )
    return await _fresh_po_response(po.id, service)


@router.get("/purchase-orders/{purchase_order_id}", response_model=PurchaseOrderResponse)
async def get_purchase_order(
    purchase_order_id: uuid.UUID,
    user: CurrentUser,
    session: DBSession,
    service: ServiceDep,
) -> PurchaseOrderResponse:
    po = await _load_visible_po(purchase_order_id, user, session, service)
    await _require_po_permission("purchase_order.read", po, user, session)
    return _po_response(po, service)


@router.patch("/purchase-orders/{purchase_order_id}", response_model=PurchaseOrderResponse)
async def update_purchase_order(
    purchase_order_id: uuid.UUID,
    payload: PurchaseOrderUpdate,
    user: CurrentUser,
    session: DBSession,
    request_ctx: RequestCtx,
    service: ServiceDep,
) -> PurchaseOrderResponse:
    po = await _load_visible_po(purchase_order_id, user, session, service)
    data = payload.model_dump(exclude_unset=True)
    expected_version = data.pop("expected_version")
    updated = await service.update_draft(
        actor=user,
        organization_id=po.organization_id,
        po_id=po.id,
        expected_version=expected_version,
        data=data,
        request_ctx=request_ctx,
    )
    return await _fresh_po_response(updated.id, service)


LifecycleCall = Callable[..., Awaitable[LifecycleResult]]


async def _run_lifecycle(
    purchase_order_id: uuid.UUID,
    permission: str,
    operation: LifecycleCall,
    user: CurrentUser,
    session: DBSession,
    service: ServiceDep,
    response: Response,
    request_ctx: RequestCtx,
    reason: str | None = None,
) -> PurchaseOrderResponse:
    po = await _load_visible_po(purchase_order_id, user, session, service)
    await _require_po_permission(permission, po, user, session)
    kwargs = {
        "actor": user,
        "organization_id": po.organization_id,
        "po_id": po.id,
        "request_ctx": request_ctx,
    }
    if reason is not None or operation.__name__ in {"withdraw", "reject", "revise", "cancel"}:
        kwargs["reason"] = reason
    result = await operation(**kwargs)
    return await _lifecycle_response(result, response, service)


@router.post("/purchase-orders/{purchase_order_id}/submit", response_model=PurchaseOrderResponse)
async def submit_purchase_order(
    purchase_order_id: uuid.UUID,
    response: Response,
    user: CurrentUser,
    session: DBSession,
    request_ctx: RequestCtx,
    service: ServiceDep,
) -> PurchaseOrderResponse:
    return await _run_lifecycle(
        purchase_order_id,
        "purchase_order.submit",
        service.submit,
        user,
        session,
        service,
        response,
        request_ctx,
    )


@router.post("/purchase-orders/{purchase_order_id}/withdraw", response_model=PurchaseOrderResponse)
async def withdraw_purchase_order(
    purchase_order_id: uuid.UUID,
    payload: LifecycleReasonPayload,
    response: Response,
    user: CurrentUser,
    session: DBSession,
    request_ctx: RequestCtx,
    service: ServiceDep,
) -> PurchaseOrderResponse:
    return await _run_lifecycle(
        purchase_order_id,
        "purchase_order.update",
        service.withdraw,
        user,
        session,
        service,
        response,
        request_ctx,
        payload.reason,
    )


@router.post("/purchase-orders/{purchase_order_id}/approve", response_model=PurchaseOrderResponse)
async def approve_purchase_order(
    purchase_order_id: uuid.UUID,
    response: Response,
    user: CurrentUser,
    session: DBSession,
    request_ctx: RequestCtx,
    service: ServiceDep,
    payload: Annotated[ApprovalPayload | None, Body()] = None,
) -> PurchaseOrderResponse:
    return await _run_lifecycle(
        purchase_order_id,
        "purchase_order.approve",
        service.approve,
        user,
        session,
        service,
        response,
        request_ctx,
        payload.reason if payload else None,
    )


@router.post("/purchase-orders/{purchase_order_id}/reject", response_model=PurchaseOrderResponse)
async def reject_purchase_order(
    purchase_order_id: uuid.UUID,
    payload: LifecycleReasonPayload,
    response: Response,
    user: CurrentUser,
    session: DBSession,
    request_ctx: RequestCtx,
    service: ServiceDep,
) -> PurchaseOrderResponse:
    return await _run_lifecycle(
        purchase_order_id,
        "purchase_order.reject",
        service.reject,
        user,
        session,
        service,
        response,
        request_ctx,
        payload.reason,
    )


@router.post("/purchase-orders/{purchase_order_id}/revise", response_model=PurchaseOrderResponse)
async def revise_purchase_order(
    purchase_order_id: uuid.UUID,
    payload: LifecycleReasonPayload,
    response: Response,
    user: CurrentUser,
    session: DBSession,
    request_ctx: RequestCtx,
    service: ServiceDep,
) -> PurchaseOrderResponse:
    return await _run_lifecycle(
        purchase_order_id,
        "purchase_order.update",
        service.revise,
        user,
        session,
        service,
        response,
        request_ctx,
        payload.reason,
    )


@router.post("/purchase-orders/{purchase_order_id}/cancel", response_model=PurchaseOrderResponse)
async def cancel_purchase_order(
    purchase_order_id: uuid.UUID,
    payload: LifecycleReasonPayload,
    response: Response,
    user: CurrentUser,
    session: DBSession,
    request_ctx: RequestCtx,
    service: ServiceDep,
) -> PurchaseOrderResponse:
    return await _run_lifecycle(
        purchase_order_id,
        "purchase_order.cancel",
        service.cancel,
        user,
        session,
        service,
        response,
        request_ctx,
        payload.reason,
    )


@router.get(
    "/purchase-orders/{purchase_order_id}/transitions",
    response_model=PurchaseOrderTransitionPage,
)
async def list_purchase_order_transitions(
    purchase_order_id: uuid.UUID,
    user: CurrentUser,
    session: DBSession,
    service: ServiceDep,
    cursor: str | None = Query(default=None),
    limit: int = Query(default=50, ge=1, le=200),
) -> PurchaseOrderTransitionPage:
    po = await _load_visible_po(purchase_order_id, user, session, service)
    await _require_po_permission("purchase_order.read", po, user, session)
    rows, next_cursor = await service.transition_repo.page_for_po(po.id, cursor=cursor, limit=limit)
    return PurchaseOrderTransitionPage(
        items=[
            PurchaseOrderTransitionResponse(
                id=row.id,
                purchase_order_id=row.purchase_order_id,
                actor_id=row.actor_id,
                from_status=row.from_status,
                to_status=row.to_status,
                operation=(row.metadata_json or {}).get("operation", ""),
                reason=row.reason,
                occurred_at=row.occurred_at,
            )
            for row in rows
        ],
        next_cursor=next_cursor,
    )
